### PR TITLE
Ingest class

### PR DESCRIPTION
**Pull request by Koncopd**
As discussed with @falexwolf this PR introduces a new Ingest class to process new small pieces of data.

> sc.pp.neighbors(adata)  # adata is huge with 1M observations
> 
> ingest = sc.Ingest(adata)  # represents the existing data, learned annotations, structure and exposes it to functionality that allows to ingest new data very quickly
> 
> adata_small.obsm['X_model'] = model(adata_small.X)
> 
> ingest.neighbors(adata_small)  # adata_small with just 1000 observations
> 
> now, we have the updated neighbors graph with 1,001,000 observations
> we want to do the same things as always
> 
> by leveraging the neighbors of the new data within the old data, 
> map the new data into the embedding (umap), by just computing a correction to the existing embedding: a new data point gets the mean position of its k nearest neighbors
> 
> ingest.umap(adata_small)
> 
> update the clustering (mapping the 1000 observations into the existing clusters): a new data point maps into a cluster if the majority of its neighbors is a member of the cluster 
> 
> ingest.louvain(adata_small)

### CODE REVIEW

**Comment by Koncopd** [comment]
This is just an initial setup.

**Comment by Koncopd** [comment]
@falexwolf 
some usage examples
https://github.com/Koncopd/anndata-scanpy-benchmarks/blob/master/use_Ingest.ipynb

File: scanpy/tools/_ingest.py
Comment (ivirshup) [inline scanpy/tools/_ingest.py]: Should this be conditional on whether the original was mean centered as well?

File: scanpy/tools/_ingest.py
Comment (ivirshup) [inline scanpy/tools/_ingest.py]: Does this take any meaningful amount of the run time? I suspect this could be made more efficient, but I'm not sure if it'd be worth it.

File: scanpy/tools/_ingest.py
Comment (ivirshup) [inline scanpy/tools/_ingest.py]: Should we be relying on pynndescent directly as in https://github.com/theislab/scanpy/pull/659?

File: scanpy/tools/_ingest.py
Comment (ivirshup) [inline scanpy/tools/_ingest.py:148]: Could this be a function (maybe in `sc.get`?) that takes an AnnData and returns its UMAP transformer?

File: scanpy/tools/_ingest.py
Comment (ivirshup) [reply scanpy/tools/_ingest.py]: Oh, also we shouldn't import UMAP at the top level. It takes a few seconds to import, so we should defer it.

**Review by ivirshup — COMMENTED** [review]
I really like that this functionality is making it into scanpy! I've comments on specific implementation details but mostly I think the methods look good. I do have a few questions about the structure, however.

* With the current functionality, could this be a module instead of a class? I'm not sure I see the advantage of this being a separate class, rather than a set of functions.
  * A potential factor here is how long it takes to run the `_init_` methods.
  * I don't think I see the `Ingest` object being updated when new data is processed. Is this correct?
* I think there should be checks to make sure the transformation is valid. Mostly this means checking the shape and var_names are compatible.

File: scanpy/tools/_ingest.py
Comment (ivirshup) [reply scanpy/tools/_ingest.py]: Thinking about this again, what's the reason to include KNN classification in here? Given the large low weight component I've seen in UMAP connectivity graphs, I'm worried this will give noisy results.

Additionally [the recent benchmarking paper](https://www.biorxiv.org/content/10.1101/644435v1) found poor relative performance for the KNN classifier they tried.

File: scanpy/tools/_ingest.py
Comment (Koncopd) [reply scanpy/tools/_ingest.py]: Yes, you are right, i just forgot about it.

File: scanpy/tools/_ingest.py
Comment (Koncopd) [reply scanpy/tools/_ingest.py]: The problem is that we don't have pynndescent as a dependency for now.
Yes, i will change to the deferred import.

**Review by Koncopd — COMMENTED** [review]
I think `map_labels` should be named `map_labels_knn`, this reflects the meaning of the function better.

**Comment by ivirshup** [comment]
> I think `map_labels` should be named `map_labels_knn`

Why change the name from `knn_classify`? I think it fits, especially since this is so close to sklearn's `KNeighborClassifier`?

**Comment by Koncopd** [comment]
> > I think `map_labels` should be named `map_labels_knn`
> 
> Why change the name from `knn_classify`? I think it fits, especially since this is so close to sklearn's `KNeighborClassifier`?

@ivirshup this was the decision of @falexwolf . Maybe something like `map_labels(..., method='knn')` which allows adding other methods when they are implemented?

**Comment by falexwolf** [comment]
Yes on your last suggestion, @Koncopd ! :)

**Comment by Koncopd** [comment]
@falexwolf @ivirshup 
New Ingest api usage
https://github.com/Koncopd/anndata-scanpy-benchmarks/blob/master/Ingest.ipynb

**Comment by LuckyMD** [comment]
This looks really cool! So when you do `ing.to_adata()` you only get back the mapped datapoints and not the whole dataset? It would be interesting to see a plot of both on top of one another to appreciate the label mapping.

**Comment by Koncopd** [comment]
@LuckyMD `ingest.to_adata` just returns a copy of `adata_new` (or the same object depending on inplace) with all mapped representations (pca, umap) and labels.

**Comment by LuckyMD** [comment]
I see. So you'd have to concatenate the data to plot them together?

**Comment by Koncopd** [comment]
Yes. Also i can add a function which returns a dataset with joint representations and labels only.

**Comment by Koncopd** [comment]
Updated the notebook above with `to_adata_joint` example - this function returns the concatenation of `adata_ref` and `adata_new` with projected representations and mapped labels.

**Comment by Koncopd** [comment]
Test are failing because of the new version of dask.

**Comment by Koncopd** [comment]
I think it is more or less complete.
Here are the tutorials
https://github.com/Koncopd/anndata-scanpy-benchmarks/blob/master/Ingest-realistic.ipynb
https://github.com/Koncopd/anndata-scanpy-benchmarks/blob/master/Ingest-simple.ipynb

**Comment by ivirshup** [comment]
Just starting to try this out, but I hit a bug. Here's a script to reproduce and the traceback:

<details>
<summary>Script </summary>

```python
import scanpy as sc
from scanpy.tools._ingest import ingest
import numpy as np
import pandas as pd
from functools import reduce

def simplify_annot(annot):
    names = annot.columns.str.extract(r"\[(.*)\].*$", expand=False)
    unique_names, idxs = np.unique(names, return_index=True)
    new_annot = annot.iloc[:, idxs].copy()
    new_annot.columns = unique_names
    return new_annot

def process(dset):
    dset.layers["counts"] = dset.X.copy()
    sc.pp.normalize_total(dset)
    sc.pp.log1p(dset)
    sc.pp.highly_variable_genes(dset)
    sc.pp.pca(dset)
    sc.pp.neighbors(dset, n_neighbors=30)
    sc.tl.umap(dset)

dset1 = sc.datasets.ebi_expression_atlas("E-GEOD-81608", filter_boring=True) 
dset2 = sc.datasets.ebi_expression_atlas("E-GEOD-83139", filter_boring=True)
# dset3 = sc.datasets.ebi_expression_atlas("E-ENAD-27", filter_boring=True)

# dsets = [dset1, dset2, dset3]
dsets = [dset1, dset2]
for dset in dsets:
    dset.obs = simplify_annot(dset.obs)
    sc.pp.calculate_qc_metrics(dset, inplace=True)

shared_genes = reduce(np.intersect1d, [dset.var_names for dset in dsets])
dsets = [dset[:, shared_genes].copy() for dset in dsets]

for dset in dsets:
    process(dset)

# dset1, dset2, dset3 = dsets
dset1, dset2 = dsets

dset1.obs["inferred cell type (dset1)"] = dset1.obs["inferred cell type"]

dset12 = ingest(dset2, dset1, obs="inferred cell type (dset1)", return_joint=True)
```

Traceback:

```python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/isaac/github/scanpy/scanpy/tools/_ingest.py", line 33, in ingest
    return ing.to_adata(inplace) if not return_joint else ing.to_adata_joint()
  File "/Users/isaac/github/scanpy/scanpy/tools/_ingest.py", line 222, in to_adata_joint
    adata = AnnData(np.vstack((self._adata_ref.X, self._adata_new.X)))
  File "/Users/isaac/github/anndata/anndata/core/anndata.py", line 566, in __init__
    filename=filename, filemode=filemode)
  File "/Users/isaac/github/anndata/anndata/core/anndata.py", line 735, in _init_as_actual
    X = X.astype(dtype, copy=False)
ValueError: setting an array element with a sequence.
```

</details>

I've reproduced on 2bffd6a1400 and after merging master with that. This occurs whether inplace is true or not. It also occurs on anndata master as well as the current release. As you'd expect, it doesn't happen with `return_joint=False`.

**Comment by Koncopd** [comment]
Thank you. I guess this is because dset{1,2}.X is sparse, i need to fix this.

**Comment by ivirshup** [comment]
Since I can get some labels out from just not getting an object back, I tried a couple things.

Using the labelled half of `E-ENAD-27` to add labels to the other cells:

<details>
<summary> plots </summary>

image

</details>

Using `E-GEOD-81608` (many cells, few labels) to label `E-GEOD-83139` (few cells, many labels) seemed to work fine considering we're moving to a lower resolution label set. Here's the latter projected onto the former:

<details>
<summary> plots </summary>

image

</details>

The reverse (former projected onto the latter) didn't seem to work so well:

<details>
<summary> plots </summary>

image

</details>

If it'd be useful, I could clean up the notebooks I ran these in and get you some confusion matrices tomorrow.

**Comment by Koncopd** [comment]
@ivirshup thanks. 
I fixed return_join, it should work now.

File: scanpy/tools/_ingest.py
Comment (ivirshup) [inline scanpy/tools/_ingest.py]: Just a few style thoughts.

* I think `_str_to_list` may be [simple enough it doesn't need a function](http://matthewrocklin.com/blog/work/2019/06/23/avoid-indirection).
* Could the preparing of `obs` and `labeling_method` (via `_str_to_list`) happen at the top of the function? I don't think it would break anything, and I find code easier to follow when a variables value is consistent through as much of the function as possible.

**Comment by ivirshup** [comment]
Great! I'll check it out when I have a chance. If this is close to ready, could it also start getting some tests?

Just to clarify, would a notebook with the pancreas integration stuff be useful to you?

**Comment by Koncopd** [comment]
Hi, yes, i will start adding tests. 
And i think it can be useful but only if you have time for it.

File: scanpy/tools/_ingest.py
Comment (Koncopd) [reply scanpy/tools/_ingest.py]: I'll fix it.

**Comment by Koncopd** [comment]
@falexwolf @ivirshup 
Tests are also ready.

**Comment by Koncopd** [comment]
Some things on pancreas dataset
https://github.com/Koncopd/anndata-scanpy-benchmarks/blob/master/ingest_pancreas.ipynb

File: scanpy/tools/_ingest.py
Comment (ivirshup) [inline scanpy/tools/_ingest.py:100]: I'm a little confused about what some of these arguments are doing and what values can be passed to them. Could you update the the docs?

File: scanpy/tools/_ingest.py
Comment (ivirshup) [inline scanpy/tools/_ingest.py:298]: What does this method do?

As it only seems to modify itself, should it be called `fit` or even something else? Based on the sklearn conventions, I would expect `transform` to return transformed values.

File: scanpy/tools/_ingest.py
Comment (ivirshup) [inline scanpy/tools/_ingest.py:134]: Do we definitely want this class to be public? At the moment, I'm not sure if there's functionality being hidden if we just expose `ingest`. If we do want to make this public, I'll want to go over it's API in more detail.

An option here is to initially merge this with the `Ingest` class being private, with the possibility of making it public in the future. What do you think?

File: scanpy/preprocessing/_simple.py
Comment (ivirshup) [inline scanpy/preprocessing/_simple.py:527]: If you write an AnnData object any values of `uns` that are `None` don't get written. While this is maybe something we should change, are you sure that all the values you rely on in `ingest` wouldn't be set to `None` with typical usage? This could be tested by checking that the results of merging into an object, and merging into a copy which has been round tripped to disk are equivalent.

File: scanpy/tools/_ingest.py
Comment (ivirshup) [inline scanpy/tools/_ingest.py:362]: Would you be interested in adding weighted knn classification? I think all we would need to do is calculate the connectivities, which I think we could do with [these lines](https://github.com/lmcinnes/umap/blob/424268f6a99a401e3b9eae9c1e8dadebf619db83/umap/umap_.py#L1828-L1837).

File: scanpy/tools/_ingest.py
Comment (ivirshup) [inline scanpy/tools/_ingest.py:437]: Could you document these functions (`to_adata` and `to_adata_joint`)? I think I know what they're supposed to do, but I'm not totally sure.

File: scanpy/tests/test_ingest.py
Comment (ivirshup) [inline scanpy/tests/test_ingest.py:65]: In general, I think we should test the functionality instead of the implementation. For example, instead of checking `assert not ing._pca_centered`, could we check if the initial PCA wasn't centered, the new data wasn't centered before it was transformed?

File: scanpy/tools/_ingest.py
Comment (ivirshup) [inline scanpy/tools/_ingest.py:295]: I don't think it does, really that class is mostly to support indexing of AnnData objects, but this class can't be subset. I suppose it could be used for the shape checks, but that's easy enough to do separately.

**Review by ivirshup — CHANGES_REQUESTED** [review]
## Notebook feedback

This looks really good! I wanted to take a closer look so I replicated it on my machine. A few plots I found useful (note I kept the original labels in `obs` as `"celltype_orig"`):

### Just show labels of added data

Here's an example:

```python
with plt.rc_context({"figure.figsize": (5, 5),"figure.dpi": 100}):
    ax = sc.pl.umap(ad_joint, show=False, size=10)
    sc.pl.umap(ad_joint[ad_joint.obs["sample"] == "Wang"], color=["celltype_orig"], ax=ax, size=10)
```

image

### Confusion matrices of changed labels

This would be very useful to see if labels change unexpectedly. Here's some code I used for this:

<details>
<summary> `confusion` function definition </summary>

```python
import numpy as np
import pandas as pd
from sklearn.metrics import confusion_matrix
import scanpy as sc

def confusion(
    orig: "Union[pd.Series, np.array, Sequence]",
    new: "Union[pd.Series, np.array, Sequence]",
    data=None,
    *,
    normalize: bool = True
) -> pd.DataFrame:
    """Given an original and new set of labels, create a labelled confusion matrix.
    
    Params
    ------
    orig
        Original labels
    new
        New labels
    normalize
        Should the confusion matrix be normalized?
        
    Usage
    -----
    >>> import seaborn as sns
    >>> confmtx = confusion(labels["orig"], labels["new"])
    >>> sns.heatmap(confmtx)
    """
    # TODO: Decide whether I want to remove this
    if isinstance(data, sc.AnnData):
        orig = data.obs[orig]
        new = data.obs[new]
    elif data is not None:
        orig = data[orig]
        new = data[new]
        
    # Coercing so I don't have to deal with it later
    orig, new = pd.Series(orig), pd.Series(new)
    assert len(orig) == len(new)

    unique_labels = pd.unique(np.concatenate((orig.values, new.values)))

    # Compute
    mtx = confusion_matrix(orig, new, labels=unique_labels)
    if normalize:
        sums = mtx.sum(axis=1)[:, np.newaxis]
        mtx = np.divide(mtx, sums, where=sums != 0)

    # Label
    if orig.name is None:
        orig_name = "Original labels"
    else:
        orig_name = orig.name
    if new.name is None:
        new_name = "New labels"
    else:
        new_name = new.name
    df = pd.DataFrame(
        mtx,
        index=pd.Index(unique_labels, name=orig_name),
        columns=pd.Index(unique_labels, name=new_name)
    )
    
    # Filter
    df = df.loc[df.index.isin(pd.unique(orig)), df.columns.isin(pd.unique(new))]
    
    return df
```


</details>

Here's an example of usage

```python
import seaborn as sns

new_labels = ad_joint.obs[ad_joint.obs["sample"] == "Wang"][["cell_type", "celltype_orig"]]
sns.heatmap(confusion(new_labels["celltype_orig"], new_labels["cell_type"]))
```

image

### Provenance

Could you include some information on where the datasets that you're integrating here came from, and how they've been processed?

### Merging larger datasets looks good!

I've merged the Segerstolpe dataset into the Baron one to test this out a bit more, and it looks really good!

<details>
<summary> Code </summary>

```python
baron = adata[adata.obs["sample"] == "Baron"].copy()
seger = adata[adata.obs["sample"] == "Segerstolpe"].copy()

sc.pp.pca(baron)
sc.pp.neighbors(baron)
sc.tl.umap(baron)

ad_bs = sc.tl.ingest(
    seger, baron,
    obs='celltype',
    embedding_method=['umap', 'pca'], k=15,
    return_joint=True
)
```

</details>

On thing that is particularly cool is all the "unclassified" cells get labeled as "schwann":

image

## `sc.tl.ingest`

### Arguments

As mentioned in review, could you write up the docs for the arguments? I just used what you have in that notebook, but it's not clear what each one does at the moment.

### `obs_names`

This might be a separate issue, but could we not change the cell labels when we join objects if there is no overlap? I'd like to be able to do: `ad_joint[ad_new.obs_names]`.

## `Ingest`

I like the state the `ingest` function is in. I think the `Ingest` class might need some more work, especially if we want to make it public. Do you definitely want `Ingest` to be made public in this PR? That could happen later.

File: scanpy/tools/_ingest.py
Comment (Koncopd) [reply scanpy/tools/_ingest.py:437]: Yes, i will add all needed descriptions soon.

File: scanpy/tools/_ingest.py
Comment (Koncopd) [reply scanpy/tools/_ingest.py:362]: Do you mean distance weighted knn?

File: scanpy/preprocessing/_simple.py
Comment (Koncopd) [reply scanpy/preprocessing/_simple.py:527]: For now they wouldn't be None, but in general it should be checked, yes.

File: scanpy/tools/_ingest.py
Comment (Koncopd) [reply scanpy/tools/_ingest.py:134]: Yes, this is the idea, to make only the function `ingest` public for now.

File: scanpy/tools/_ingest.py
Comment (Koncopd) [reply scanpy/tools/_ingest.py:100]: Definitely i will do it.

File: scanpy/tools/_ingest.py
Comment (Koncopd) [reply scanpy/tools/_ingest.py:298]: It does transform `adata_new` to the same representation which was used by `adata_ref` to calculate neighbors. But it doesn't return any values, yes, maybe fit is a better name for this function

File: scanpy/tests/test_ingest.py
Comment (Koncopd) [reply scanpy/tests/test_ingest.py:65]: Do we need testing that the initial pca wasn't centered here?
I mean it is specified as `sc.pp.pca(adata_ref, n_comps=30, zero_center=False)` a few lines above, so it should be centered and i don't see any need to test it here.

File: scanpy/tools/_ingest.py
Comment (Koncopd) [reply scanpy/tools/_ingest.py:295]: I found it to be convenient here, maybe i need to get rid of it because it is an internal AnnData thing.

**Comment by Koncopd** [comment]
@ivirshup 
Thank you for your analysis.
I will definitely add all needed arguments' descriptions a bit later.

As for the `Ingest` class, i don't think it should be public, the public thing is the `ingest` function.

Also i added the possibility of specifying `adata.concatenate` arguments in both `ingest` and `to_adata_joint`. So to get the same `obs_names` you can use `sc.tl.ingest(..., index_unique=None)` or `ing.to_adata_joint(index_unique=None)`.

File: scanpy/tools/_ingest.py
Comment (ivirshup) [reply scanpy/tools/_ingest.py:362]: Something like that, I was thinking UMAP’s connectivity.

File: scanpy/tests/test_ingest.py
Comment (ivirshup) [reply scanpy/tests/test_ingest.py:65]: When you check the value of `_pca_centered` you’re checking some internal state, not what computation happened. I think this should be tested more like this (this is basically pseudo code, since I’m not at a computer):

```python
refc = sc.pp.pca(adata, zero_center=True, copy=True)
ref = sc.pp.pca(adata, zero_center=False, copy=True)

# Not sure if I have these calls exactly right
# The intent is the `new` objects have been projected into the ref objects pca space
newc = sc.tl.ingest(adata, refc, embedding_method=“pca”, return_joint=False)
new = sc.tl.ingest(adata, ref, embedding_method=“pca”, return_joint=False)

assert np.allclose(newc.obsm[“X_pca”], refc.obsm[“X_pca”])
assert np.allclose(new.obsm[“X_pca”], ref.obsm[“X_pca”])
assert not np.allclose(new.obsm[“X_pca”], newc.obsm[“X_pca”])
```

You could alternatively use an oracle here, like scikit-learn’s PCA transformer.

Checking the value of `_pca_centered` doesn’t tell us whether these cases are true. Additionally, testing the behavior of the method rather than internal construction means we’re free to change the internals without having to rewrite tests.

This is basically the [“test behavior, not implementation”](https://testing.googleblog.com/2013/08/testing-on-toilet-test-behavior-not.html) mantra.

**Comment by flying-sheep** [comment]
@Koncopd once this is ready from your side, you can re-request @ivirshup’s review.

**Comment by falexwolf** [comment]
I'd be fine with it, but happy to wait for @ivirshup's review!

**Comment by flying-sheep** [comment]
Well, he reviewed it before and none of the conversations is marked as resolved. Let’s let them close those and then we can squash & merge.

File: scanpy/datasets/__init__.py
Comment (flying-sheep) [inline scanpy/datasets/__init__.py]:
```suggestion
    adata = read(filename, backup_url=url)
```

**Review by flying-sheep — CHANGES_REQUESTED** [review]
Why all the string quote changes? Black should read its config and leave those alone.

Also there’s a merge error, please pay real good attention that things like this don’t happen.

**Review by flying-sheep — APPROVED** [review]

**Comment by Koncopd** [comment]
@flying-sheep Yeah, sorry, missed this. Thank you for correcting.

As for black, i don't understand why it happened. I changed the quotes back and run black on the file, now it doesn't change it...

I will go through the discussion with @ivirshup again, fix some things and then will re-request the review.

**Comment by flying-sheep** [comment]
Great, thank you a lot, this seems like it was quite some work!

I think the resolve feature is pretty effective in keeping track of what one already adressed.

**Comment by ivirshup** [comment]
@Koncopd, I'm happy to go over this again whenever. I haven't looked too closely yet, but it looks like a lot has been addressed.

**Comment by falexwolf** [comment]
Guys, I'll merge this for now so that I can conveniently play around with it in practical settings and potentially improve. We don't need to advertise for now and the next release might be a bit ahead anyway.

I'm reading the latest comments as you being essentially positive. ☺️

**Pull request merged**
